lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -0,0 +1,688 @@
+                with the same name as provided via `relation_name` argument.
+            RelationInterfaceMismatchError: The relation with the same name as provided
+                via `relation_name` argument does not have the `prometheus_scrape` relation
+                interface.
+            RelationRoleMismatchError: If the relation with the same name as provided
+                via `relation_name` argument does not have the `RelationRole.requires`
+                role.
+        """
+        _validate_relation_by_interface_and_direction(
+            charm, relation_name, RELATION_INTERFACE_NAME, RelationRole.provides
+        )
+
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._endpoint_schema = endpoint_schema
+        self._endpoint_address = endpoint_address
+        self._endpoint_port = int(endpoint_port)
+        self._endpoint_path = endpoint_path
+
+        on_relation = self._charm.on[self._relation_name]
+        self.framework.observe(
+            on_relation.relation_created,
+            self._set_endpoint_on_relation_change,
+        )
+        self.framework.observe(
+            on_relation.relation_joined,
+            self._set_endpoint_on_relation_change,
+        )
+
+    def update_endpoint(self, relation: Optional[Relation] = None) -> None:
+        """Triggers programmatically the update of the relation data.
+
+        This method should be used when the charm relying on this library needs
+        to update the relation data in response to something occurring outside
+        of the `prometheus_remote_write` relation lifecycle, e.g., in case of an
+        host address change because the charmed operator becomes connected to an
+        Ingress after the `prometheus_remote_write` relation is established.
+
+        Args:
+            relation: An optional instance of `class:ops.model.Relation` to update.
+                If not provided, all instances of the `prometheus_remote_write`
+                relation are updated.
+        """
+        relations = [relation] if relation else self.model.relations[self._relation_name]
+
+        for relation in relations:
+            self._set_endpoint_on_relation(relation)
+
+    def _set_endpoint_on_relation_change(self, event: RelationEvent) -> None:
+        self._set_endpoint_on_relation(event.relation)
+
+    def _set_endpoint_on_relation(self, relation: Relation) -> None:
+        """Set the the remote_write endpoint on relations.
+
+        Args:
+            relation: Optional relation. If provided, only this relation will be
+                updated. Otherwise, all instances of the `prometheus_remote_write`
+                relation managed by this `PrometheusRemoteWriteProvider` will be
+                updated.
+        """
+        address = self._endpoint_address or self._get_relation_bind_address()
```

Comment:
I wonder if there needs to be a check for having obtained a valid address here. I mean are we guaranteed that a bind address will always be available for a relation at any state of the lifecycle of the charm.